Add crc64 support

deadalnix opened the pull request:
As per title.

deadalnix commented:
This is not quite ready yet, but as I can't run the tests manually because of https://issues.dlang.org/show_bug.cgi?id=17107 , I have to rely on autotester.

deadalnix commented:
Alright this is in better shape now. Tests are green? No idea what circle is unhappy about.

JackStouffer commented:
cc @wilzbach

deadalnix commented:
Also, among interesting questions, CRC64 has 2 standards: ECMA and OSI. The algorithm is the same, but the polynomial is different ( 0xD800000000000000 for OSI, 0xC96C5795D7870F42 for ECMA ). ECMA is generally considered better, but maybe we'd like to support both ?

deadalnix commented:
Ok Added support for both CRC64-ECMA and CRC64-ISO, that wasn't significantly harder, and can be leveraged to do CRC32C and other variations as well.

deadalnix commented:
I have no idea why circle is complaining. That'd be great if someone would look at this is less than a week.

JackStouffer commented:
There seems to be a dearth of reviews in Phobos right now.

Do you know of anyone who would be qualified to review this code?

deadalnix commented:
@JackStouffer I asked someone who dug into this CRC thing with me to figure it out. He'll be able to assert the validitity of the algorithm, but is not an experimented D coder, so that'd be great to have someone checks that.

Also, I'd be glad if someone could explain me what circle is complaining about.

JackStouffer commented:
Also, needs a changelog entry

wilzbach commented on std/digest/crc.d:
```diff
@@ -169,36 +197,46 @@ struct CRC32
         unittest
         {
-            CRC32 dig;
+            CRC dig;
```
That's the reason why CircleCi fails with:

```d
out/std_digest_crc.d(46): Error: struct std.digest.crc.CRC(uint N, ulong P) if (N == 32 || N == 64) is used as a type
out/std_digest_crc.d(59): Error: struct std.digest.crc.CRC(uint N, ulong P) if (N == 32 || N == 64) is used as a type
out/std_digest_crc.d(71): Error: struct std.digest.crc.CRC(uint N, ulong P) if (N == 32 || N == 64) is used as a type
out/std_digest_crc.d(73): Error: undefined identifier 'R'
```

The script at CircleCi seperates all public unittests and tries to run them with the newly compiled release library to ensure that all unittests on dlang.org are runnable as they are.

Btw I remember there was a long discussion last summer about the problems of using `unittest` blocks in templated code - in short in theory it's discouraged at Phobos because the `unittest` block gets instantiated very every new instance.

deadalnix replied:
Indeed, the test is generated for each templates, it is a 3 line tests, so I don't think there is really any problem here.

deadalnix commented:
Sadly I find myself unable to continue this work.

https://issues.dlang.org/show_bug.cgi?id=17236

I was already required to jump though hoops to run the tests, but this is getting out of hands.

wilzbach commented:
> Sadly I find myself unable to continue this work.
> I was already required to jump though hoops to run the tests, but this is getting out of hands.

As [mentioned on the NG](http://forum.dlang.org/post/fqkpbcsldzveevuanjak@forum.dlang.org) I am still not sure what your exact problem is. 
In any case, you can always build a module directly if you have a recent DMD compiler, e.g.:

```
source $(curl https://i.dlang.io | bash -s dmd-nightly -a)
```

deadalnix commented:
Added doc for N, P and R.

Rebased. I got various merge conflicts for style changes that arguably do not improve the code in any way. Especially the one which puts template constraint on their own line.

JackStouffer commented on std/digest/crc.d:
```diff
@@ -355,6 +502,42 @@ ubyte[4] crc32Of(T...)(T data)
+ *      CRC64-ECMA of data
+ */
+//simple alias doesn't work here, hope this gets inlined...
+ubyte[8] crc64ECMAOf(T...)(T data)
```
According to Circle, this needs an example

deadalnix replied:
I'm not sure what I'm expected to do here. I looked at crc32Of and there is no example. Is there a piece of code I can look at to know what to do ?

JackStouffer commented on std/digest/crc.d:
```diff
@@ -355,6 +502,42 @@ ubyte[4] crc32Of(T...)(T data)
+ *      CRC64-ISO of data
+ */
+//simple alias doesn't work here, hope this gets inlined...
+ubyte[8] crc64ISOOf(T...)(T data)
```
This too

deadalnix replied:
this too too !

JackStouffer commented:
@deadalnix Was that person you mentioned able to verify the CRC code correctness?

deadalnix commented:
@ftrader , we need you :)

deadalnix commented:
You can check it matches the test vector from go here: https://golang.org/src/hash/crc64/crc64_test.go

JackStouffer commented:
@deadalnix Sorry, I only just saw your comment about verification.

Any chance you can fix the public example issue so this can be pulled?

CyberShadow commented:
FWIW, I had a CRC64 (ECMA variant) implementation on top of std.digest lying around. I plugged the unittests from here into my implementation, and everything worked.

deadalnix commented:
OK @CyberShadow confirm that the result are good, which is pretty convincing the algorithm is correct (this is a hash function after all, so you'd expect to get very different results).

So I rebased and I want to get that in good shape so we can merge it.

deadalnix commented:
Looks like circle is happy. Do we move forward with that one ?

JackStouffer commented on std/digest/crc.d:
```diff
@@ -138,11 +135,53 @@ private static immutable uint[256][8] crc32Tables = genTables(0xEDB88320);
+         * Type of the finished CRC hash.
+         * ubyte[8] if N is 32, ubyte[16] if N is 64.
+         */
+        alias R = ubyte[T.sizeof];
```
Shouldn't `R` be `ubyte[4]` for 32 bit? Because `finish` used to return `ubyte[4]` originally and now it returns `ubyte[8]`.

deadalnix replied:
The code is correct, but the comment is wrong. Fixing.

deadalnix commented:
Fixed.
